ingestion-beam/src/main/java/com/mozilla/telemetry/Republisher.java

```diff
@@ -0,0 +1,143 @@
+    decoded //
+        .apply(Deduplicate.markAsSeen(options.getParsedRedisUri(),
+            options.getDeduplicateExpireSeconds()))
+        .errorsTo(errorCollections);
+
+    // Republish debug messages.
+    if (options.getEnableDebugDestination()) {
+      RepublisherOptions.Parsed opts = options.as(RepublisherOptions.Parsed.class);
+      opts.setOutput(options.getDebugDestination());
+      decoded //
+          .apply("FilterDebugMessages", Filter.by(message -> {
+            message = PubsubConstraints.ensureNonNull(message);
+            return message.getAttribute("x_debug_id") != null;
+          })) //
+          .apply(opts.getOutputType().write(opts));
+    }
+
+    // Republish to per-docType destinations.
+    final List<String> enabledDocTypes = Optional //
+        .ofNullable(options.getPerDocTypeEnabledList()) //
+        .orElse(Collections.emptyList());
+    for (String entry : enabledDocTypes) {
+      final String[] components = entry.split("/");
+      final String targetNamespace;
+      final String targetDocType;
+      if (components.length == 1) {
+        targetNamespace = "telemetry";
+        targetDocType = components[0];
+      } else {
+        targetNamespace = components[0];
+        targetDocType = components[1];
+      }
+      RepublisherOptions.Parsed opts = options.as(RepublisherOptions.Parsed.class);
+      opts.setOutput(NestedValueProvider.of(options.getPerDocTypeDestination(),
+          pattern -> pattern.replace("${document_namespace}", targetNamespace)
+              .replace("${document_type}", targetDocType)));
+      decoded //
+          .apply("DocType is " + targetDocType, Filter.by(message -> {
+            message = PubsubConstraints.ensureNonNull(message);
+            return targetNamespace.equals(message.getAttribute("document_namespace"))
+                && targetDocType.equals(message.getAttribute("document_type"));
+          })) //
+          .apply(opts.getOutputType().write(opts));
+    }
+
+    // Republish to sampled per-channel destinations.
+    final Map<String, Double> ratioMap = Optional //
+        .ofNullable(options.getPerChannelSampleRatios()) //
+        .orElse(new HashMap<>());
+    for (Map.Entry<String, Double> entry : ratioMap.entrySet()) {
+      String targetChannel = entry.getKey();
+      Double ratio = entry.getValue();
+
+      RepublisherOptions.Parsed opts = options.as(RepublisherOptions.Parsed.class);
+      opts.setOutput(NestedValueProvider.of(options.getPerChannelDestination(),
+          pattern -> pattern.replace("${channel}", targetChannel)));
+
+      decoded //
+          .apply("RandomlySample" + StringUtils.capitalize(targetChannel), Filter.by(message -> {
+            message = PubsubConstraints.ensureNonNull(message);
+            String channel = message.getAttribute("app_update_channel");
+            return targetChannel.equals(channel)
+                && ThreadLocalRandom.current().nextDouble() < ratio;
```

Comment:
I believe this should be randomly sampling by `clientId` rather than uniformly across all pings.

Reply:
Breaking this off to https://github.com/mozilla/gcp-ingestion/issues/519